[torch][segment_reduce] Support for multi dimension (cpu only)

serhaty opened the pull request:
Summary:
Add support for multi-d input for cpu forward/backward implementation.

Next step: Adding cuda support for multi-d input.

Test Plan: Added unit tests.

Differential Revision: D29105457

ngimel commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -29,52 +29,71 @@ Tensor _segment_reduce_cpu_kernel(
-          }
+        int64_t lengths_cum_sum = 0;
+        for (int64_t i = 0; i < segment_count; ++i) {
+          for (int64_t l = 0; l < stride_count; ++l) {
```
why are you splitting stride_count and seg_element_count into 2 loops? They are conceptually the same, the dimensions that are not being reduced.

ngimel reviewed (commented):
This code is becoming quite complicated, and yet it can only handle reduction over 0-th axis for a contiguous tensor, and cannot easily be extended to handle other axes. At this point, the implementation should be switched to TensorIterator. Please see e.g. how tensorIterator is handling `max` along dimension https://github.com/pytorch/pytorch/blob/8e92a3a8b026178d65fb995e2a4b750fea93a093/aten/src/ATen/native/cpu/TensorCompareKernel.cpp#L20-L125. The idea behind this is you create a tensor iterator squashing the dimension of interest, and let TensorIterator iterate over all other dimensions, while providing a loop function to handle what you want to do along the dimension of interest.  It will be a bit more complicated for segmented reduction instead of full reduction, but the advantage is that there's already a good structure that allows templated operations across arbitrary dimensions, and you are utilizing TensorIterator for efficient iteration over tensor, not putting in any requirements that tensor is contiguous.

serhaty commented on aten/src/ATen/native/SegmentReduce.cpp:
```diff
@@ -29,52 +29,71 @@ Tensor _segment_reduce_cpu_kernel(
-  auto output = at::empty({batch_size}, data.options());
+  int64_t segment_count = lengths.numel();
+  auto output_shape = data.sizes().vec();
+  output_shape[0] = segment_count;
```
good catch, I am actually checking this on caller side.

Commit:
[torch][segment_reduce] Add cuda support for mean reduction

Differential Revision: D28922838

fbshipit-source-id: 6544b91df1ed2bc4ef50191c9016395023f3e2ea

Commit:
[torch][segment_reduce] Support for multi dimension (cpu only) (#59951)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/59951

Add support for multi-d input for cpu forward/backward implementation.

Next step: Adding cuda support for multi-d input.

Test Plan: Added unit tests.

Differential Revision: D29105457

fbshipit-source-id: d61fe767a80410501272231219d751cf29225b0b
